radar-auth/src/main/java/org/radarcns/auth/token/AbstractRadarToken.java

```diff
@@ -0,0 +1,112 @@
+ * permission checks.
+ */
+public abstract class AbstractRadarToken implements RadarToken {
+
+    protected static final String CLIENT_CREDENTIALS = "client_credentials";
+
+    @Override
+    public boolean hasPermission(Permission permission) {
+        return hasScope(permission.scopeName())
+                && (isClientCredentials() || hasAuthorityForPermission(permission));
+
+    }
+
+    @Override
+    public boolean hasPermissionOnProject(Permission permission, String projectName) {
+        return hasScope(permission.scopeName())
+                && (isClientCredentials() || hasAuthorityForPermission(permission, projectName));
+    }
+
+    @Override
+    public boolean hasPermissionOnSubject(Permission permission, String projectName,
+            String subjectName) {
+        return hasScope(permission.scopeName())
+                && (isClientCredentials() || hasAuthorityForPermission(permission, projectName,
+                        subjectName));
+    }
+
+    protected boolean hasScope(String scope) {
+        return getScopes().contains(scope);
+    }
+
+    protected boolean isClientCredentials() {
+        return CLIENT_CREDENTIALS.equals(getGrantType());
+    }
+
+    /**
+     * Check all authorities in this token, project and non-project specific, for the given
+     * permission.
+     * @param permission the permission
+     * @return {@code true} if any authority contains the permission, {@code false} otherwise
+     */
+    protected boolean hasAuthorityForPermission(Permission permission) {
+        return getRoles().values().stream()
+                .flatMap(s -> s.stream())
+                .anyMatch(permission::isAuthorityAllowed)
+                || hasNonProjectRelatedAuthorityForPermission(permission);
+    }
+
+    /**
+     * Check authorities in this token linked to the given project, or not linked to any project
+     * (such as {@code SYS_ADMIN}), for the given permission.
+     * @param permission the permission
+     * @param projectName the project name
+     * @return {@code true} if any authority contains the permission, {@code false} otherwise
+     */
+    protected boolean hasAuthorityForPermission(Permission permission, String projectName) {
+        return getRoles().containsKey(projectName) && getRoles().get(projectName).stream()
+                .anyMatch(permission::isAuthorityAllowed)
+                || hasNonProjectRelatedAuthorityForPermission(permission);
+    }
+
+    /**
+     * Check authorities in this token linked to the given project, or not linked to any project
+     * (such as {@code SYS_ADMIN}), for the given permission on the given subject.
+     * @param permission the permission
+     * @param projectName the project name
+     * @param subjectName the subject name
+     * @return {@code true} if any authority contains the permission, {@code false} otherwise
+     */
+    protected boolean hasAuthorityForPermission(Permission permission, String projectName,
```

Comment:
Looks good to me. Minor comment. Would it help to have different method names than method overloading for these methods. e.g. `hasAuthorityForPermissionInProject` ?

Reply:
Indeed had to choose between long method names and overloading :stuck_out_tongue: probably a longer method name is still better.